<a href="https://colab.research.google.com/github/jasmineA20/MathModeling/blob/main/HW%238.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import random
random.seed(12)

Section 5.1 #3 

approximating pi by considering the number of random points selected inside the quarter circle Q: $x^2 + y^2 = 1, x \geq{0}, y \geq 0$ where the quarter circle is taken to be inside the square \\
 S: 0 $\leq x \leq 1$ and 0 $\leq y \leq 1$

Input: total number of random points to be generated: n \\
Output: Estimated value of pi \\

1) Initialize counter = 0 \\
2) for i = 1,2,..N do steps 3-5 \\
3) Generate random numbers for x_i and y_i such that 0<=x_i<=1 and 0<=y_i<=1\\
4) If random y_i <= f(x_i, y_i) increment counter by 1 \\
5) leave loop and calculate area value  counter/nval \\
6) Multiply the area by 4 since this was only a quarter of a circle to get the approximate value of pi. \\

In [5]:
counter = 0

n = [100, 200, 300, 400, 500, 1000, 10000, 100000 ]
for n_val in n: 
  counter = 0 
  for j in range(n_val):
    x,y = random.random(), random.random()
    Q = x**2 + y**2 
    if Q <= 1: 
      counter += 1
  area_val = counter/n_val 
  approx_pi = 4 * area_val
  print("Pi approximation when n is {} = ".format(n_val),approx_pi)

Pi approximation when n is 100 =  3.04
Pi approximation when n is 200 =  3.06
Pi approximation when n is 300 =  3.2
Pi approximation when n is 400 =  3.11
Pi approximation when n is 500 =  3.128
Pi approximation when n is 1000 =  3.152
Pi approximation when n is 10000 =  3.1544
Pi approximation when n is 100000 =  3.14296


Section 5.1 #5

Area trapped between two curves $ y = x^2$ and y = 6 - x and the x and y axes 

Input: Total number n_val of random points to be generated in the simulation. \\
Output: area between the given curves. \\
1) Initialize counter = 0 \\
2) for i = 1,2,...,N do step 3-5 \\
3) Generate random numbers x,y such that it satisfies both curves. \\
4) If the value is between the range then increment counter by 1 \\
5) Calculate the area between the two curves. \\
6) Calculate counter \\
7) Coutput area = 6*2*counter/n_val

In [4]:
n = [100, 200, 300, 400, 500, 1000, 10000, 100000 ]
for n_val in n: 
  counter = 0 
  for i in range(n_val):
    x,y = 2*random.random(), 6*random.random()
    y1 = x*x
    y2 = 6-x

    if y > y1 and y < y2: 
      counter += 1

  area = 6*2*counter/n_val
  print("Approximation of area when n is {} = ".format(n_val),area)

Approximation of area when n is 100 =  6.96
Approximation of area when n is 200 =  7.14
Approximation of area when n is 300 =  7.2
Approximation of area when n is 400 =  7.44
Approximation of area when n is 500 =  7.392
Approximation of area when n is 1000 =  6.936
Approximation of area when n is 10000 =  7.4148
Approximation of area when n is 100000 =  7.3296


Section 5.3 #2

input: Total number n of 5 coin flips in the simulation. \\
output: The probability of three heads occuring. \\
1) Inititalize counter to 0. \\
2) iterate over the number of flips generated and repeat steps 3 and 4 
. \\
3) Assuming heads occured for x<=0.5, increment counter for this condition. \\
4) Check if counter is 3. If yes, break if not, go back to 2. \\
5) Compute the probabiltiy as number of points for which heads occured and divide by number of flips. \\
6) Print the probability computed. 

In [6]:
counter = 0
n = [100, 200, 300, 400, 500, 1000, 10000, 100000 ]
for n_val in n: 
  counter = 0 
  for i in range(n_val):
    count = 0
    for _ in range(5):
      x = random.random()
      if x<= .5:
        count += 1
    if count == 3: 
      counter +=1 

  probability = counter/n_val
  print("The probability when n is {} = ".format(n_val),probability)

The probability when n is 100 =  0.39
The probability when n is 200 =  0.31
The probability when n is 300 =  0.31666666666666665
The probability when n is 400 =  0.34
The probability when n is 500 =  0.294
The probability when n is 1000 =  0.309
The probability when n is 10000 =  0.3102
The probability when n is 100000 =  0.31302


mean score for throwing five darts for n = 1000 from above is 0.309 

Section 5.3 #4

(i) \\
0 <= x_1i <= 0.1, roll1 = 1 \\
0.1 <= x_1i <= 0.2, roll1 = 2 \\
0.2 <= x_1i <= 0.4, roll1 = 3 \\
0.4 <= x_1i <= 0.7, roll1 = 4 \\
0.7 <= x_1i <= 0.9, roll1 = 5 \\
0.9 <= x_1i <= 1, roll1 = 6 \\

(ii) \\
0 <= x_2i <= 0.3, roll2 = 1 \\
0.3 <= x_2i <= 0.4, roll2 = 2 \\
0.4 <= x_2i <= 0.6, roll2 = 3 \\
0.6 <= x_2i <= 0.7, roll2 = 4 \\
0.7 <= x_2i <= 0.75, roll2 = 5 \\
0.75 <= x_2i <= 1, roll2 = 6 \\

1) Initialize counters to 0 \\
2) for all i from 1 to n, n = 300 \\
3) find two numbers such that 0<= x_1i < 1 and 0<= x_2i <= 1. \\
4) if x_i1 belongs to [0,1], set roll1 to values as described by (i) \\
5) if x_2i belongs to [0,1], set roll2 to values as described by (ii) \\
6) Calculate sum roll1+roll2 \\
7) Increment counter 
8) calculate probability of each sum i = 0,1,...,12 = counter/n

In [7]:
n = 300 

counter = np.zeros([13,1])

for i in range(n):
  x1,x2 = random.random(), random.random()

  if x1 <= .1:
    roll_1 = 1
  elif x1 <=.2:
    roll_1 = 2
  elif x1 <=.4:
    roll_1 =3
  elif x1 <=.7:
    roll_1 = 4
  elif x1 <=.9:
    roll_1 = 5
  else: 
    roll_1 = 6

  if x2 <= .3:
    roll_2 = 1
  elif x2 <= .4:
    roll_2 = 2     
  elif x2 <= .6:
    roll_2 = 3
  elif x2 <= .7:
    roll_2 = 4
  elif x2 <= .75:
    roll_2 = 5
  else: 
    roll_2 = 6

  sum_rolls = roll_1 + roll_2 
  counter[sum_rolls] = counter[sum_rolls]+1

probabilities = counter/n 
print('Probabilites when n = 300 are: ')   
print(probabilities)   

Probabilites when n = 300 are: 
[[0.        ]
 [0.        ]
 [0.02666667]
 [0.02666667]
 [0.08666667]
 [0.13666667]
 [0.16      ]
 [0.15      ]
 [0.12333333]
 [0.11333333]
 [0.08      ]
 [0.06666667]
 [0.03      ]]


Final Problem

Two fair coins flipped simultaneously and Player 1 calls "Evens" or "Odds". "Evens" means that coins must match (both heads or both tails), and "Odds" means that coins will not match. If PLayer 1 calls correctly then he/she wins $1 from Player 2. If player 1 calls incorrectly then Player 2 wins $1 from Player 1. 

Coins flips yield {HH, HT, TH, TT} with probabilities  \\
P(HH) = 1/4 (Even) \\
P(HT) = P(TH) = 2/4 (Odd) \\
P(TT) = 1/4 (Even) \\
P(Even) = 2/4 = 1/2 \\
P(Odd) = 2/4 = 1/2 \\
Expected payoff = (1*(1/2) + (-1)*(1/2))*n = 0 \\

With either strategy, the expected payoff is still $0 since the probability of winning and losing is equal. 


